### ISSUE TITLE
VoiceOver duplicates alerts in iframes

### ISSUE THREAD

**Issue by terracoda** [phetsims/a11y-research#132]
I always seem to hear ARIA Live alerts twice when using a screen reader with the A11y View. I do not hear `aria-valuetext` twice.

@zepumph, mentioned yesterday that there is a `role="presentation"` on the copy of the PDOM to silence it. Perhaps we need the same on the copy of the `aria-live-regions` from where the alerts are delivered?

Not a huge priority, but this would greatly improve the A11y View if someone happens to be using a screen reader. This would be good to fix before our next conference where a demonstration might happen.

Assigning to @jessegreenberg for re-assigning and prioritizing.

**Comment by jessegreenberg** [comment]
Interesting, I am not sure why alerts are getting duplicate alerts sometimes. @terracoda I assume you are using VoiceOver right? I don't get any duplicate alerts when running the a11y view in Firefox with NVDA. We are not copying any of the aria-live elements into the a11y-view. Ill try adding `role="presentation"` on the yellow aria-live box to see if that helps.

**Comment by jessegreenberg** [comment]
role='presentation' has been added to the list in the a11y view showing alerts. @terracoda can you please test to see if that fixed the issue?

**Comment by terracoda** [comment]
@jessegreenberg, this might only happen with sliders. I was testing Gravity Force Lab Basics when I logged the issue. In GFL:B I regular hear duplicate force alerts on focus and when moving.

I just checked John Travoltage's Leg Swing slider and the electron alert repeats pretty regularly in that sim, too, when moving the leg and picking up changes. 

In Resistance and a Wire the duplicate alerts are much less frequent. I only seem to get a duplicate alert for the changing letters and changing resistance after moving one of the sliders quickly and with more than 3 quick key presses. 

Hmm, I do get duplicate alerts in Friction, too, several alerts repeat. They do not repeat when I use the regular sim.

Yes, I am using Safari and VoiceOver on Mac OS 10.13.6

**Comment by jessegreenberg** [comment]
OK, thanks @terracoda, Ill take a look at the sliders specificially in GFL:B/RIAW/Friction.

**Comment by terracoda** [comment]
@jessegreenberg as you know, Friction does not have a slider, but the alerts are repeating. I haven't tested BASE.

**Comment by jessegreenberg** [comment]
OK sounds good. We are trying a service called [MacinCloud](https://www.macincloud.com/) for debugging since I don't have access to a Mac. Ill take a look there, and if VoiceOver works on that service I will be able to debug this fully.

**Comment by jessegreenberg** [comment]
I tried MacinCloud today, I wasn't able to get audio to work on it so I couldn't hear VO. I submitted a report to them, hopefully we will hear back soon. VO seems to work on that service, I can see the black focus highlight move around, but the display is stuck on 
![image](https://user-images.githubusercontent.com/6396244/53130395-1fd4ef00-3538-11e9-951a-aec5b113c43e.png)

So I can't see what it is saying. Ill try again when we hear back from them.

**Comment by terracoda** [comment]
@jessegreenberg, perhaps the repeat alerts when using VoiceOver has to do with the iframe? I have long suspected this as I also hear a lot of duplication in the sonification wrapper when I use Voice Over.

In the latest sonified [sonified GFL:Basics](https://phet-dev.colorado.edu/html/phet-io-wrapper-sonification/1.0.0-dev.101/phet-io-wrapper-sonification/gravity-force-lab-basics/html/gravity-force-lab-basics-sonification.html?sonificationFile=sonificationOptions&massSound=0&forceSound=3&massBoundaryLimitSound=2&accessibility)

I just heard repeat UI alerts for all the checkboxes and repeat force alerts when moving the spheres. 
I also heard repeat force alerts when changing Mass, but only if I change Mass quickly.

However, FYI, I cleared my cache and reloaded the sim and now things are not repeating.

This is very strange behaviour. Let's not spend too much time on this since you don't ever hear repeat alerts in NVDA. It seems to be a VoiceOver problem, and nobody but us uses the A11y View and the sonification wrapper.

I'll keep some notes on when I hear repeat alerts when testing.

**Comment by jessegreenberg** [comment]
That is a good idea @terracoda, interesting that it is happening in the sonification wrappers too.

> However, FYI, I just cleared my cache and reloaded the sim and now things are not repeating.

That is also good to know. Removing my assignment for now, will try again once I have access to a test device. I would like to try an aria-live alert in a very basic iframe to see if we can get repetition outside of  a sim.

**Comment by emily-phet** [comment]
@jessegreenberg 

@terracoda mentioned this repeating behavior occurred in an interview with a consultant yesterday - not in the a11y-view, but in the sonification wrapper. Would it be worth asking the QA team to spend some time trying to reproduce this repeating behavior in contexts where the sim is in an iframe? It seems like it would be good to track this down, since we often user test with wrappers and repeating behavior can derail an interview.

**Comment by jessegreenberg** [comment]
OK, that sounds good. @KatieWoe could you please test this link with macOS VoiceOver and see if you get duplicate alerts?

https://phet-dev.colorado.edu/html/gravity-force-lab-basics/1.0.0-dev.38/phet/gravity-force-lab-basics_a11y_view.html

In https://github.com/phetsims/a11y-research/issues/132#issuecomment-466956319 @terracoda mentioned that clearing cache and reloading made the bug go away, so if you don't see the bug please reload the sim a couple of times and try again.

Then, can you then try to reproduce in this simple iframe example and let me know if you can? Again, try reloading if it doesn't happen right away.

https://phet-dev.colorado.edu/html/jg-tests/aria-live-iframe/iframe-example.html

Thank you!

**Comment by KatieWoe** [comment]
First link it did repeat alerts.  Second link, the screen reader would read instructions when you focus on the button.  If you push it during these instructions, it reads the alert twice.  If you do not, it does not.

**Comment by jessegreenberg** [comment]
Thanks @KatieWoe, sounds good. @twant sorry to ask again, but this is another VoiceOver specific problem so would you be able to help debug this? Could we meet again next week to take a look? This one should be easier to reproduce :)

**Comment by twant** [comment]
@jessegreenberg Yep, let's take a look next week! I'll slack you about times. 

@KatieWoe does this happen on all browsers and versions? I know Jesse and I tested for a different bug on a couple different versions of Safari (and got different results on each), as well as Chrome. Let me know if there's a specific browser you'd like me to use as Jesse and I work on this!

**Comment by KatieWoe** [comment]
The platforms we support and test on are:
- VO on Latest MacOS and latest safari
- VO on latest iOS iPad
- NVDA on Latest Windows 10 and latest firefox
- Jaws on Latest Windows 10 and latest chrome

**Comment by jessegreenberg** [comment]
Since I have access to a macOS now I shouldn't need to pair for this. Thanks for the offer to help @twant, Ill let you know if I need assistance!

I just saw this in another JSFiddle while looking into another issue.

**Comment by jessegreenberg** [comment]
I just verified that this is still an issue.

**Comment by jessegreenberg** [comment]
Changing the content HTML in https://github.com/phetsims/a11y-research/issues/132#issuecomment-527689724 to use a slider instead of a button has the alerts duplicate every time.
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>contentpage</title>
  </head>
  <body>
    <button id="button">Trigger alert</button>

    <input type="range" id="slider">
    <p id="alerter" aria-live="polite">Alert here...</p>
  </body>

  <script type="text/javascript">
    const button = document.getElementById( 'button' );
    const slider = document.getElementById( 'slider' );
    const alerter = document.getElementById( 'alerter' );

    let counter = 0;
    button.addEventListener( 'slider', ( event ) => {
      alerter.textContent = '';
      alerter.hidden = false;
      alerter.textContent = 'New Alert: ' + counter++;

      window.setTimeout( () => {
        // alerter.textContent = '';
        alerter.hidden = true;
      }, 200 );
    } );

    slider.addEventListener(  'change', ( event ) => {
      alerter.textContent = 'Slider Alert: ' + counter++;
    } );

  </script>
</html>
```

So the duplication happens when there is some input, then announced value change, then aria-live alert.

**Comment by jessegreenberg** [comment]
I searched for any info online or other reports of this and didn't see anything. I also tried different combinations of the `aria-live` attributes to see if any would for some reason prevent this VO bug but nothing made an impact. I don't think this is anything we can work around unfortunately. I will write a bug report and submit to apple.

Pinging @terracoda and @emily-phet so they are aware of the status of this issue for user testing.

**Comment by jessegreenberg** [comment]
Report generated but Feedback Assistant is temporarily out of service. Will try again later.

**Comment by emily-phet** [comment]
@jessegreenberg Thanks. Sounds like we'll have to live with repeated alerts when user testing or designing with people using screen readers. This is another reason to move the sound design work away from using the wrappers, which I think is where we're headed.

Haptics and other work still in wrappers will just need to keep this in mind when planning studies, etc.

**Comment by terracoda** [comment]
@jessegreenberg, indeed the duplication is inconsistent, even in the example links you provided above. Apologies for being late to this issue. For me this duplication happens in the current A11y Views for GFLB in the following cases:
- checking and unchecking checkboxes (consistantly)
- reset button (never)

**Comment by terracoda** [comment]
I added the above comment because I wasn't sure a checkbox was a value change, but maybe it is? It certainly is in some cases.

**Comment by jessegreenberg** [comment]
OK thanks @terracoda that aligns with what I was seeing. Maybe instead of value it is more just any state change that gets read by VO like checked, pressed, value or things like that.

**Comment by jessegreenberg** [comment]
For this issue, we should see if it is still a problem and consider submitting a report to webkit (instead of Feedback assistant).

**Comment by marlitas** [comment]
@jessegreenberg will submit the report to webkit for this when he is submitting a report for https://github.com/phetsims/a11y-research/issues/198

**Comment by jessegreenberg** [comment]
This has already been reported to webkit bugzilla. I added to the thread.

**Comment by lisnykh** [comment]
Hello guys, sorry any updates regarding this issue? I still see duplicate alerts when reading web content with VoiceOver and it is placed withing iframe

**Comment by jessegreenberg** [comment]
Hi @lisnykh! No updates from our side yet. This is a WebKit/VoiceOver issue, and others are noticing the same duplicate alerts in iframes. For example:

https://bugs.webkit.org/show_bug.cgi?id=294253
https://bugs.webkit.org/show_bug.cgi?id=299134

We’ll monitor those bugs and update here if anything changes.
